[AOTInductor] Double buffering for Weights

muchulee8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114446

Summary:
This adds function to model container doing weight swapping with double buffering.

There are 2 parts for double buffering
a) Write constants into inactive buffer
b) Swap active buffer

For (a), we write the constants into the buffer that's currently not in use, and store the information in both constants map and the corresponding constant array to read.
For (b), we obtain the lock, and activate the constant map/constant array that is inactive, and flag the one that's currently in use to inactive.

Test Plan:
test/cpp/aot_inductor/test.cpp

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @aakhundov @ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

muchulee8 committed:
[WIP] [AOTInductor] Test for double buffering

Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

muchulee8 commented:
@muchulee8 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51543732).

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[WIP] [AOTInductor] Test for double buffering"


Summary:
NA

Test Plan:
NA

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 committed:
Update on "[AOTInductor] Double buffering for Weights"


Summary:
This adds function to model container doing weight swapping with double buffering.

Test Plan:
test/cpp/aot_inductor/test.cpp

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 commented on torch/csrc/inductor/aoti_runtime/model.h:
```diff
@@ -379,22 +390,43 @@ class AOTInductorModelBase {
+      constants_ = std::make_shared<std::vector<AtenTensorHandle>>(
+          constants_info_.size());
+    } else {
+      constants_->resize(constants_info_.size());
```
Yes. They are unchanged. So this is basically a no-op after constants get initiated.

This is actually just copying the original code (previously, `update_constants_map` does (1) update the constants map (2) Populate the corresponding tensor in constant map to the array.) Now I have moved (2) to this separate function (`update_constants_array_from_map`.)

chenyang78 replied:
OK, I see. So, the else-branch is executed when we first initialize the ```constants``` vector.

muchulee8 commented on torch/csrc/inductor/aoti_runtime/model_container.h:
```diff
@@ -89,6 +95,101 @@ class AOTInductorModelContainer {
+      aoti_torch_get_data_ptr(it->second, &user_constant_ptr);
+      aoti_torch_get_storage_size(it->second, &constant_size);
+
+      AOTI_RUNTIME_DEVICE_CHECK(cudaMemcpy(
```
The latter question, whether it's on device or on host are both possible, and there's no trivial way to identify whether it's on device or on host.

For the former question, I did a copy for safeness. I don't think we can "steal the pointer", the underlying storage for the tensor is a raw pointer, unlike smart pointers, stealing the ownership afaik is not possible. We can definitely just "use the pointer", I don't see issue for this. But for such case, we also don't manage the data, and something might break if the data managed by the user get freed.

eellison replied:
> The latter question, whether it's on device or on host are both possible, and there's no trivial way to identify whether it's on device or on host.

I feel like this should be possible...

> don't think we can "steal the pointer", the underlying storage for the tensor is a raw pointer, unlike smart pointer

If the input is an PyTorch tensor, the way you do this is by setting it's deallocation function function to be a no-op, and then manage our own deallocation function which would call into the caching allocation and free it's pointer.  See - code changing dealloc function [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/cuda/Module.cpp#L1099-L1107). 

Although, there might be some abi complications here, and we could do as a follow up.

chenyang78 reviewed (approved):
LGTM. Thanks!

eellison reviewed (commented):
would you please update the summary with details of the pr ?

eellison commented on torch/csrc/inductor/aoti_runtime/model_container.h:
```diff
@@ -89,6 +95,101 @@ class AOTInductorModelContainer {
+
+  void swap_constant_buffer() {
+    std::lock_guard unique_lk(model_exec_mutex_);
+    // Need to wait?
```
What do you mean by this `// Need to wait?`

eellison commented on test/cpp/aot_inductor/test.py:
```diff
@@ -30,10 +30,13 @@ def forward(self, x, y):
         f"model_so_path_{device}": model_so_path,
         f"inputs_{device}": [x, y],
         f"outputs_{device}": [ref_output],
+        f"fc_weight_{device}": params["fc.weight"],
+        f"fc_bias_{device}": params["fc.bias"],
```
Why was this necessary ?

eellison commented on torch/csrc/inductor/aoti_runtime/model_container.h:
```diff
@@ -89,6 +95,101 @@ class AOTInductorModelContainer {
+      int64_t* stride;
+      int64_t offset;
+      int device_idx = -1;
+      AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_get_strides(it->second, &stride));
```
Maybe you're already doing this but add asserts offsets/strides are the same ?

eellison reviewed (approved):
Looks good

muchulee8 commented on torch/csrc/inductor/aoti_runtime/model_container.h:
```diff
@@ -126,12 +227,25 @@ class AOTInductorModelContainer {
+  std::vector<size_t> constants_internal_offset_;
 #endif // USE_CUDA
 
+  bool use_secondary_;
```
Added.

muchulee8 committed:
Update on "[AOTInductor] Double buffering for Weights"


Summary:
This adds function to model container doing weight swapping with double buffering.

There are 2 parts for double buffering
a) Write constants into inactive buffer
b) Swap active buffer

For (a), we write the constants into the buffer that's currently not in use, and store the information in both constants map and the corresponding constant array to read.
For (b), we obtain the lock, and activate the constant map/constant array that is inactive, and flag the one that's currently in use to inactive.

Test Plan:
test/cpp/aot_inductor/test.cpp

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng aakhundov ColinPeppler

Differential Revision: [D51543732](https://our.internmc.facebook.com/intern/diff/D51543732)

[ghstack-poisoned]

muchulee8 commented:
@pytorchbot merge
